File: C++/Tree Alogrithms/balanced_binary_tree_check.cpp
#include <bits/stdc++.h>
using namespace std;


/*
	Worst case time complexity: O(n)
	Auxiliary space complexity: O(h); where h = height of binary tree
*/


struct Node {
	int data;
	Node *left;
	Node *right;

	Node(int value) {
		data = value;
		left = NULL;
		right = NULL;
	}
};


int isBalanced(Node *root) {
	if (root == NULL) return 0;

	int leftHeight = isBalanced(root->left);
	if (leftHeight == -1) return -1;

	int rightHeight = isBalanced(root->right);
	if (rightHeight == -1) return -1;

	if (abs(leftHeight - rightHeight) > 1) return -1;
	else return 1 + max(leftHeight, rightHeight);
}


int main() {
	Node *root = new Node(10);
	// left
	root->left = new Node(20);
	root->left->left = new Node(30);
	root->left->right = new Node(40);
	// right
	root->right = new Node(50);
	root->right->right = new Node(60);
	
	cout << "Entered tree is balanced? "
		<< (isBalanced(root) != -1 ? "true" : "false")
		<< "\n";

	return 0;
}

File: Python/Tree Algorithms/balanced_binary_tree_check.py
# Python program for implementation of Balanced Binary Tree Check

class Node: 
    def __init__(self, data): 
        self.data = data 
        self.left = None
        self.right = None
        
# compute tree's height
def height(root): 
    if root is None: 
        return 0
    return max(height(root.left), height(root.right)) + 1
  
def isBalanced(root): 
# For a None tree balance
    if root is None: 
        return True
    
# compute left root height
    left_height = height(root.left) 
# compute right root height
    right_height = height(root.right) 
    
# if the left height and right height of tree is more than 1, then the tree is not balanced
    if (abs(leftheight - rightheight) <= 1) and isBalanced(root.left) is True 
    
# if the right subtree is not balanced, this tree is not balanced then also     
    and isBalanced( root.right) is True: 
        return True
    return False
  
root = Node(1) 
root.left = Node(2) 
root.right = Node(3) 
root.left.left = Node(4) 
root.left.right = Node(5) 
root.left.left.left = Node(8) 

print("Is the given tree balanced? ")
result = "True" if isBalanced(root) else "False"
print(result)
